reports/filesDataExporter/batch/batch.go

```diff
@@ -0,0 +1,171 @@
+	}
+)
+
+const exportSeconds metrics.ID = "exportSeconds"
+
+const labelFilesCollection = "filesCollection"
+
+func (s *batchDataExporter) Export(ctx context.Context, lastSuccessfulRun time.Time) error {
+	bucketsRateLimit := make(chan bool, s.bucketsRateLimit)
+
+	buckets, err := s.handlers.ListSourceBuckets(ctx)
+	if err != nil {
+		s.logger.Error().Err(err).Msg("failed to list source buckets")
+		return errors.Wrap(err, "failed to list source buckets")
+	}
+
+	ch := make(chan *exportError)
+	for _, b := range buckets {
+		go s.exportBucket(ctx, lastSuccessfulRun, b.Name, ch, bucketsRateLimit)
+	}
+
+	var errCount int
+	for i := 0; i < len(buckets); i++ {
+		syncErr := <-ch
+		if syncErr != nil {
+			s.logger.Error().Err(syncErr.err).Str("bucket", syncErr.id).Msg("failed to sync")
+			errCount++
+		}
+	}
+
+	if errCount > 0 {
+		s.logger.Error().Msgf("%d failure(s) out of %d bucket(s) to export", errCount, len(buckets))
+		return errors.Errorf("%d failure(s) out of %d bucket(s) to export", errCount, len(buckets))
+	}
+
+	return nil
+}
+
+// GetPrometheusMetricsCollection returns all prometheus metrics collectors to be registered
+func (s *batchDataExporter) GetPrometheusMetricsCollection() map[metrics.ID]prometheus.Collector {
+	return s.metricsCollection
+}
+
+func New(handlers filesDataExporter.Handlers, bucketsRateLimit int, logger zerolog.Logger) filesDataExporter.BatchFilesDataExporter {
+	logger = logger.With().Str("component", "reports/filesDataExporter/batch").Logger()
+
+	metricsCollection := make(map[metrics.ID]prometheus.Collector)
+	h := prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "batch",
+		Name:      "file_export_seconds",
+		Help:      "Time taken to export file",
+	}, []string{"operation", "success", "result"})
+	metricsCollection[exportSeconds] = h
+
+	return &batchDataExporter{
+		handlers:          handlers,
+		bucketsRateLimit:  bucketsRateLimit,
+		logger:            logger,
+		metricsCollection: metricsCollection,
+	}
+}
+
+func (s *batchDataExporter) exportBucket(ctx context.Context, lastSuccessfulRun time.Time, bucketID string, errCh chan *exportError, rateLimit chan bool) {
+	rateLimit <- true
+
+	files, err := s.handlers.ListSourceFilesAsc(ctx, bucketID, strfmt.DateTime(lastSuccessfulRun))
+	if err != nil {
+		s.logger.Error().Err(err).Str("bucket", bucketID).Msg("failed to list source files")
+		errCh <- &exportError{bucketID, errors.Wrap(err, fmt.Sprintf("failed to list source files in bucket %s", bucketID))}
+		return
+	}
+
+	var exportCount int
+	var errCount int
+
+	for _, f := range files {
+		select {
+		case <-ctx.Done():
+			s.logger.Error().Str("bucket", bucketID).Msg("aborting bucket export due to context cancellation")
+			errCh <- &exportError{bucketID, errors.Wrap(ctx.Err(), fmt.Sprintf("aborting bucket export due to context cancellation"))}
+			return
+		default:
+			if !utils.SliceContains(f.Labels, labelFilesCollection) && time.Time(f.Created).After(lastSuccessfulRun) {
```

Comment:
invalid operation: f (variable of type *invalid type) has no field or method Labels